File: erettsegi_emelt/2017_okt/c#/Hianyzasok.cs
using System;
using System.IO;
using System.Collections.Generic;

var napnev = new []{ "vasarnap", "hetfo", "kedd", "szerda", "csutortok", "pentek", "szombat" };
var napszam = new []{ 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 335 };

var hianyzasok = new List<Hianyzas>();
var honapNapSzamlalok = new int[2];

foreach(var line in File.ReadAllLines("naplo.txt")) {
    var split = line.Split(' ');

    if(split[0] == "#") {
        honapNapSzamlalok[0] = int.Parse(split[1]);
        honapNapSzamlalok[1] = int.Parse(split[2]);
    }else{
        hianyzasok.Add(new Hianyzas(split[0] + ' ' +  split[1], split[2], honapNapSzamlalok[0], honapNapSzamlalok[1]));
    }
}

Console.WriteLine($"2. Feladat: Hiányzások száma: {hianyzasok.Count}");

var igazoltHianyzasok = 0;
var igazolatlanHianyzasok = 0;

foreach(var hiany in hianyzasok) {
    for(var i = 0; i < hiany.orak.Length; ++i) {
        var ora = hiany.orak[i];

        if(ora == 'X') {
            ++igazoltHianyzasok;
        }else if(ora == 'I') {
            ++igazolatlanHianyzasok;
        }
    }
}

Console.WriteLine($"3. Feladat: Igazolt hiányzások: {igazoltHianyzasok}, igazolatlanok: {igazolatlanHianyzasok}");
Console.WriteLine("5. Feladat: Írjon be egy hónapot és egy napot");

var beHonap = int.Parse(Console.ReadLine());
var beNap = int.Parse(Console.ReadLine());

Console.WriteLine($"Azon a napon: {Hetnapja(beHonap, beNap)} volt");
Console.WriteLine("6. Feladat: Írja be 1 nap nevét és 1 óraszámot");

var beTanNap = Console.ReadLine();
var beOraszam = int.Parse(Console.ReadLine()) - 1;
var hianyzottakSzama = 0;

foreach(var hiany in hianyzasok) {
    if(beTanNap == Hetnapja(hiany.honap, hiany.nap)) {
        var ora = hiany.orak[beOraszam];

        if(ora == 'X' || ora == 'I') {
            ++hianyzottakSzama;
        }
    }
}

Console.WriteLine($"Ekkor {hianyzottakSzama}-an hiányoztak");
Console.WriteLine("7. Feladat: ");

var hianyzasMap = new Dictionary<string, int>();
foreach(var hiany in hianyzasok) {
    var hianyzottakk = 0;

    for(var i = 0; i < hiany.orak.Length; ++i) {
        var ora = hiany.orak[i];

        if(ora == 'X' || ora == 'I') {
            ++hianyzottakk;
        }
    }

    hianyzasMap.TryGetValue(hiany.nev, out int eddigiHianyzottak);
    hianyzasMap.Add(hiany.nev, eddigiHianyzottak + hianyzottakk);
}

var legtobbHianyzas = -1;
foreach(var v in hianyzasMap.Values) {
    if(v > legtobbHianyzas) {
        legtobbHianyzas = v;
    }
}

foreach(var (k, v) in hianyzasMap) {
    if(v == legtobbHianyzas) {
        Console.Write(k + ' ');
    }
}

string Hetnapja(int honap, int nap) => napnev[(napszam[honap - 1] + nap) % 7];

File: erettsegi/emelt/2017_okt/java/Hianyzasok.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Hianyzasok {

    public static void main(String[] args) throws IOException {
        var hianyzasok = new ArrayList<Hianyzas>();
        var honap = 0;
        var nap = 0;

        for(var line : Files.readAllLines(Path.of("naplo.txt"))) {
            var split = line.split(" ");

            if(split[0].equals("#")) {
                honap = Integer.parseInt(split[1]);
                nap = Integer.parseInt(split[2]);
            }else{
                hianyzasok.add(new Hianyzas(split[0] + ' ' +  split[1], split[2], honap, nap));
            }
        }

        System.out.println("2. Feladat: Hiányzások száma: " + hianyzasok.size());

        var igazoltHianyzasok = 0;
        var igazolatlanHianyzasok = 0;

        for(var hiany : hianyzasok) {
            for(var i = 0; i < hiany.orak.length(); ++i) {
                var ora = hiany.orak.charAt(i);

                if(ora == 'X') {
                    ++igazoltHianyzasok;
                }else if(ora == 'I') {
                    ++igazolatlanHianyzasok;
                }
            }
        }

        System.out.println("3. Feladat: Igazolt hiányzások: " + igazoltHianyzasok + ", igazolatlanok: " + igazolatlanHianyzasok);
        System.out.println("5. Feladat:");

        var beHonap = Integer.parseInt(System.console().readLine("Írj be egy hónap számot: "));
        var beNap = Integer.parseInt(System.console().readLine("Írj be egy nap számot: "));

        System.out.println("Azon a napon: " + hetnapja(beHonap, beNap) + " volt");
        System.out.println("6. Feladat:");

        var beTanNap = System.console().readLine("Írj be egy nap nevet! ");
        var beOraszam = Integer.parseInt(System.console().readLine("Írj be egy óraszámot! ")) - 1;
        var hianyzottak = 0;

        for(var hiany : hianyzasok) {
            if(beTanNap.equals(hetnapja(hiany.honap, hiany.nap))) {
                var ora = hiany.orak.charAt(beOraszam);

                if(ora == 'X' || ora == 'I') {
                    ++hianyzottak;
                }
            }
        }

        System.out.println("Ekkor " + hianyzottak + "-an hiányoztak");
        System.out.println("7. Feladat: ");

        var hianyzasMap = new HashMap<String, Integer>();
        for(var hiany : hianyzasok) {
            var hianyzottak = 0;

            for(var i = 0; i < hiany.orak.length(); ++i) {
                var ora = hiany.orak.charAt(i);

                if(ora == 'X' || ora == 'I') {
                    ++hianyzottak;
                }
            }

            hianyzasMap.put(hiany.nev, hianyzasMap.getOrDefault(hiany.nev, 0) + hianyzottak);
        }

        var legtobbHianyzas = -1;
        for(var v : hianyzasMap.values()) {
            if(v > legtobbHianyzas) {
                legtobbHianyzas = v;
            }
        }

        for(var e : hianyzasMap.entrySet()) {
            if(e.getValue() == legtobbHianyzas) {
                System.out.print(e.getKey() + ' ');
            }
        }
    }

    public static final String[] napnev = { "vasarnap", "hetfo", "kedd", "szerda", "csutortok", "pentek", "szombat" };
    public static final int[] napszam = { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 335 };

    public static String hetnapja(int honap, int nap) {
        return napnev[(napszam[honap - 1] + nap) % 7];
    }
}